Propagate the error string of GIF processing for decode_gif

yongtang opened the pull request:
This fix tries to improve the error thrown by `decode_gif` to include the error string generated by GIF processing.

Previously, the error was not very indicative as the error string
returned by GIF processing was hidden:
```
..........
InvalidArgumentError (see above for traceback): Invalid GIF data, size 2091369
	 [[Node: DecodeGif = DecodeGif[_device="/job:localhost/replica:0/task:0/device:CPU:0"](ReadFile)]]
```

This fix propagate the error string (`can't process optimized gif`) to be part of the `InvalidArgumentError`:
```
InvalidArgumentError (see above for traceback): Invalid GIF data (size 2091369), can't process optimized gif
         [[Node: DecodeGif = DecodeGif[_device="/job:localhost/replica:0/task:0/device:CPU:0"](ReadFile)]]
```

This fix fixes #15838.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Propagate the error message to exception of `InvalidArgumentError` for `decode_gif`

This fix tries to improve the error thrown by `decode_gif`
to include the error string generated by GIF processing.

Previously, the error was not very indicative as the error string
returned by GIF processing was hidden:
```
..........
InvalidArgumentError (see above for traceback): Invalid GIF data, size 2091369
	 [[Node: DecodeGif = DecodeGif[_device="/job:localhost/replica:0/task:0/device:CPU:0"](ReadFile)]]
```

This fix propagate the error string to be part of the `InvalidArgumentError`:
```
InvalidArgumentError (see above for traceback): Invalid GIF data (size 2091369), can't process optimized gif
         [[Node: DecodeGif = DecodeGif[_device="/job:localhost/replica:0/task:0/device:CPU:0"](ReadFile)]]
```

This fix fixes 15838.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Reformat with clang-format -i --style=Google

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

rmlarsen commented on tensorflow/core/lib/gif/gif_io.h:
```diff
@@ -43,7 +43,8 @@ namespace tensorflow {
 
 uint8* Decode(const void* srcdata, int datasize,
-              std::function<uint8*(int, int, int, int)> allocate_output);
+              std::function<uint8*(int, int, int, int)> allocate_output,
```
Please add a unit test for the new functionality.

yongtang committed:
Add unit test for decode_gif to cover changes

so that error messages like `can't process optimized gif` are
propagated to the exception of `InvalidArgumentError`.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @rmlarsen for the review. A unit test `testOptimizedGifErrorString` has been added to image_ops_test.py. Please take a look and let me know if there are any issues.
